File: android/example/src/main/java/io/github/ivere27/volvoxgrid/example/HierarchyJsonDemo.kt
package io.github.ivere27.volvoxgrid.example

import io.github.ivere27.volvoxgrid.Align
import io.github.ivere27.volvoxgrid.CellInteraction
import io.github.ivere27.volvoxgrid.CellStyle
import io.github.ivere27.volvoxgrid.ColIndicatorCellMode
import io.github.ivere27.volvoxgrid.ColIndicatorCellModes
import io.github.ivere27.volvoxgrid.ColIndicatorConfig
import io.github.ivere27.volvoxgrid.ColumnDataType
import io.github.ivere27.volvoxgrid.ColumnDef
import io.github.ivere27.volvoxgrid.CornerIndicatorConfig
import io.github.ivere27.volvoxgrid.CornerIndicatorSlot
import io.github.ivere27.volvoxgrid.CornerIndicatorSlotKind
import io.github.ivere27.volvoxgrid.DefineColumnsRequest
import io.github.ivere27.volvoxgrid.EditActivation
import io.github.ivere27.volvoxgrid.EditConfig
import io.github.ivere27.volvoxgrid.EditTrigger
import io.github.ivere27.volvoxgrid.Font
import io.github.ivere27.volvoxgrid.GridConfig
import io.github.ivere27.volvoxgrid.HeaderFeatures
import io.github.ivere27.volvoxgrid.IndicatorsConfig
import io.github.ivere27.volvoxgrid.IndicatorAppearance
import io.github.ivere27.volvoxgrid.InteractionConfig
import io.github.ivere27.volvoxgrid.LayoutConfig
import io.github.ivere27.volvoxgrid.LoadDataOptions
import io.github.ivere27.volvoxgrid.LoadDataStatus
import io.github.ivere27.volvoxgrid.OutlineConfig
import io.github.ivere27.volvoxgrid.ResizePolicy
import io.github.ivere27.volvoxgrid.RowIndicatorConfig
import io.github.ivere27.volvoxgrid.RowIndicatorSlot
import io.github.ivere27.volvoxgrid.RowIndicatorSlotKind
import io.github.ivere27.volvoxgrid.SelectionConfig
import io.github.ivere27.volvoxgrid.SelectionMode
import io.github.ivere27.volvoxgrid.ThemePreset
import io.github.ivere27.volvoxgrid.TreeIndicatorStyle
import io.github.ivere27.volvoxgrid.VolvoxGridController
import org.json.JSONArray
import org.json.JSONObject

object HierarchyJsonDemo {
    private const val NAME_COLUMN_INDEX = 0
    private const val SIZE_COLUMN_INDEX = 2
    private const val MODIFIED_COLUMN_INDEX = 3
    private const val PERMISSIONS_COLUMN_INDEX = 4
    const val ACTION_COLUMN_INDEX = 5

    private const val SHORT_DATE_FORMAT = "short date"

    private val captions = arrayOf("Name", "Type", "Size", "Modified", "Permissions", "Action")
    private val keys = arrayOf("Name", "Type", "Size", "Modified", "Permissions", "Action")
    private const val TREE_COLOR = 0xFFA8A29E.toInt()
    private const val FOLDER_TEXT_COLOR = 0xFF92400E.toInt()
    private const val ACTION_TEXT_COLOR = 0xFF2563EB.toInt()
    private const val OUTLINE_INDENT = 20
    private const val MIN_OUTLINE_INDICATOR_WIDTH = 56
    private const val NAME_EXPANDER_WIDTH = 280
    private const val HEADER_BAND_ROWS = 1
    private const val HEADER_ROW_HEIGHT = 44

    fun load(controller: VolvoxGridController) {
        val rawJson = controller.getDemoData("hierarchy").toString(Charsets.UTF_8)
        val rows = JSONArray(rawJson)
        val levels = hierarchyLevels(rows)
        val types = List(rows.length()) { row -> rows.getJSONObject(row).optString("Type") }
        val sanitizedJson = visibleRowsJson(rows).toString()
        controller.setColCount(keys.size)
        controller.defineColumns(hierarchyColumnRequest())
        val result = controller.loadData(
            sanitizedJson.toByteArray(Charsets.UTF_8),
            LoadDataOptions.newBuilder()
                .setAutoCreateColumns(false)
                .build()
        )
        check(result.status != LoadDataStatus.LOAD_FAILED) { "LoadData failed for embedded hierarchy demo" }
        controller.configure(hierarchyThemeConfig(maxOutlineDepth(levels), maxOutlineLevel(levels)))

        val actionStyle = CellStyle.newBuilder()
            .setForeground(ACTION_TEXT_COLOR)
            .build()
        val folderStyle = CellStyle.newBuilder()
            .setForeground(FOLDER_TEXT_COLOR)
            .setFont(Font.newBuilder().setBold(true).build())
            .build()

        for (row in levels.indices) {
            val isFolder = row < types.size && types[row] == "Folder"
            controller.setRowOutlineLevel(row, levels[row])
            controller.setCellStyleRange(
                row,
                ACTION_COLUMN_INDEX,
                row,
                ACTION_COLUMN_INDEX,
                actionStyle
            )
            if (isFolder) {
                controller.setCellStyleRange(row, NAME_COLUMN_INDEX, row, NAME_COLUMN_INDEX, folderStyle)
            }
        }
        controller.autoSize(0, keys.lastIndex)
    }

    private fun hierarchyColumnRequest(): DefineColumnsRequest {
        val builder = DefineColumnsRequest.newBuilder()
        for (col in keys.indices) {
            val def = ColumnDef.newBuilder()
                .setIndex(col)
                .setCaption(captions[col])
                .setKey(keys[col])
            when (col) {
                SIZE_COLUMN_INDEX -> def.align = Align.ALIGN_RIGHT_CENTER
                MODIFIED_COLUMN_INDEX -> {
                    def.dataType = ColumnDataType.COLUMN_DATA_DATE
                    def.format = SHORT_DATE_FORMAT
                }
                PERMISSIONS_COLUMN_INDEX, ACTION_COLUMN_INDEX -> def.align = Align.ALIGN_CENTER_CENTER
            }
            if (col == ACTION_COLUMN_INDEX) {
                def.interaction = CellInteraction.CELL_INTERACTION_TEXT_LINK
            }
            if (col == NAME_COLUMN_INDEX) {
                def.hidden = true
            }
            builder.addColumns(def.build())
        }
        return builder.build()
    }

    private fun visibleRowsJson(rows: JSONArray): JSONArray {
        val visibleRows = JSONArray()
        for (row in 0 until rows.length()) {
            val source = rows.getJSONObject(row)
            val target = JSONObject()
            for (key in keys) {
                target.put(key, source.opt(key) ?: "")
            }
            visibleRows.put(target)
        }
        return visibleRows
    }

    private fun hierarchyLevels(rows: JSONArray): List<Int> {
        val rowsById = mutableMapOf<String, JSONObject>()
        for (row in 0 until rows.length()) {
            val source = rows.getJSONObject(row)
            val id = source.optString("Id")
            check(id.isNotBlank()) { "Hierarchy demo row is missing Id" }
            rowsById[id] = source
        }

        val cache = mutableMapOf<String, Int>()
        fun depthOf(row: JSONObject, visiting: MutableSet<String>): Int {
            val id = row.optString("Id")
            cache[id]?.let { return it }
            check(visiting.add(id)) { "Hierarchy demo data contains a parent cycle at $id" }
            val parentValue = row.opt("ParentId")
            val parentId = if (parentValue == null || parentValue == JSONObject.NULL) "" else parentValue.toString()
            val depth = if (parentId.isBlank()) {
                0
            } else {
                val parent = rowsById[parentId]
                    ?: error("Hierarchy demo data references missing parent $parentId")
                depthOf(parent, visiting) + 1
            }
            visiting.remove(id)
            cache[id] = depth
            return depth
        }

        return List(rows.length()) { row -> depthOf(rows.getJSONObject(row), mutableSetOf()) }
    }

    private fun maxOutlineDepth(levels: List<Int>): Int {
        val nonNegativeLevels = levels.filter { it >= 0 }
        val minLevel = nonNegativeLevels.minOrNull() ?: 0
        val maxLevel = levels.maxOrNull() ?: 0
        return (maxLevel - minLevel).coerceAtLeast(0)
    }

    private fun maxOutlineLevel(levels: List<Int>): Int {
        return levels.filter { it >= 0 }.maxOrNull() ?: 0
    }

    private fun outlineIndicatorWidth(maxOutlineDepth: Int): Int {
        return maxOf(MIN_OUTLINE_INDICATOR_WIDTH, (maxOutlineDepth.coerceAtLeast(0) + 1) * OUTLINE_INDENT)
    }

    private fun expanderIndicatorWidth(maxOutlineDepth: Int): Int {
        return outlineIndicatorWidth(maxOutlineDepth) + NAME_EXPANDER_WIDTH
    }

    private fun hierarchyThemeConfig(maxOutlineDepth: Int, maxOutlineLevel: Int): GridConfig {
        val outlineWidth = outlineIndicatorWidth(maxOutlineDepth)
        val expanderWidth = expanderIndicatorWidth(maxOutlineDepth)
        return GridConfig.newBuilder()
            .setThemePreset(ThemePreset.THEME_AMBER)
            .setLayout(
                LayoutConfig.newBuilder()
                    .setFixedRows(0)
                    .build()
            )
            .setSelection(
                SelectionConfig.newBuilder()
                    .setMode(SelectionMode.SELECTION_FREE)
                    .build()
            )
            .setEditing(
                EditConfig.newBuilder()
                    .setActivation(EditActivation.newBuilder()
                        .setTrigger(EditTrigger.EDIT_TRIGGER_NONE)
                        .build())
                    .build()
            )
            .setOutline(
                OutlineConfig.newBuilder()
                    .setTreeIndicator(TreeIndicatorStyle.TREE_INDICATOR_ARROWS_LEAF)
                    .setIndicatorIndent(OUTLINE_INDENT)
                    .setMaxLevels(maxOutlineLevel.coerceAtLeast(0))
                    .setShowLevelButtons(true)
                    .setLabelColumn(NAME_COLUMN_INDEX)
                    .setTreeColor(TREE_COLOR)
                    .build()
            )
            .setInteraction(
                InteractionConfig.newBuilder()
                    .setResize(
                        ResizePolicy.newBuilder()
                            .setColumns(true)
                            .setRows(false)
                            .build()
                    )
                    .setAutoSizeMouse(true)
                    .setHeaderFeatures(
                        HeaderFeatures.newBuilder()
                            .setSort(false)
                            .setReorder(false)
                            .setChooser(false)
                            .build()
                    )
                    .build()
            )
            .setIndicators(
                IndicatorsConfig.newBuilder()
                    .setRowStart(
                        RowIndicatorConfig.newBuilder()
                            .setVisible(true)
                            .setWidth(expanderWidth)
                            .setAutoSize(false)
                            .setAllowResize(true)
                            .addSlots(
                                RowIndicatorSlot.newBuilder()
                                    .setKind(RowIndicatorSlotKind.ROW_INDICATOR_SLOT_EXPANDER)
                                    .setWidth(expanderWidth)
                                    .setVisible(true)
                                    .build()
                            )
                            .build()
                    )
                    .setCornerTopStart(
                        CornerIndicatorConfig.newBuilder()
                            .setVisible(true)
                            .addSlots(
                                CornerIndicatorSlot.newBuilder()
                                    .setKind(CornerIndicatorSlotKind.CORNER_SLOT_OUTLINE_LEVELS)
                                    .setWidth(outlineWidth)
                                    .setVisible(true)
                                    .build()
                            )
                            .build()
                    )
                    .setColTop(
                        ColIndicatorConfig.newBuilder()
                            .setVisible(true)
                            .setDefaultRowHeight(HEADER_ROW_HEIGHT)
                            .setBandRows(HEADER_BAND_ROWS)
                            .setCellModes(
                                ColIndicatorCellModes.newBuilder()
                                    .addModes(ColIndicatorCellMode.COL_INDICATOR_CELL_HEADER_TEXT)
                                    .build()
                            )
                            .setAllowResize(true)
                            .build()
                    )
                    .setAppearance(IndicatorAppearance.INDICATOR_APPEARANCE_MODERN)
                    .build()
            )
            .build()
    }
}

File: flutter/example/lib/hierarchy_json_demo.dart
import 'dart:convert';
import 'dart:typed_data';

import 'package:flutter/foundation.dart'
    show TargetPlatform, defaultTargetPlatform;
import 'package:volvoxgrid/volvoxgrid.dart' hide Padding;

const int _hierNameColumnWidth = 260;
const int _hierTypeColumnWidth = 80;
const int _hierSizeColumnWidth = 80;
const int _hierModifiedColumnWidth = 120;
const int _hierPermissionsColumnWidth = 100;
const int _hierActionColumnWidth = 92;
const List<int> _hierarchyColWidths = [
  _hierNameColumnWidth,
  _hierTypeColumnWidth,
  _hierSizeColumnWidth,
  _hierModifiedColumnWidth,
  _hierPermissionsColumnWidth,
  _hierActionColumnWidth,
];
const List<String> _hierarchyCaptions = [
  'Name',
  'Type',
  'Size',
  'Modified',
  'Permissions',
  'Action',
];
const List<String> _hierarchyKeys = [
  'Name',
  'Type',
  'Size',
  'Modified',
  'Permissions',
  'Action',
];
const int hierarchyNameColumn = 0;
const int _hierarchySizeColumn = 2;
const int _hierarchyModifiedColumn = 3;
const int _hierarchyPermissionsColumn = 4;
const int hierarchyActionColumn = 5;
const int _hierTreeColor = 0xFFA8A29E;
const int _hierFolderTextColor = 0xFF92400E;
const int _hierActionTextColor = 0xFF2563EB;
const int _hierOutlineIndent = 20;
const int _hierMinOutlineIndicatorWidth = 56;
const int _hierNameExpanderWidth = 280;
const int _hierHeaderBandRows = 1;
const int _hierDesktopHeaderHeight = 28;
const int _hierMobileHeaderHeight = 44;
const String _hierShortDateFormat = 'short date';

bool get _hierTouchHeader {
  switch (defaultTargetPlatform) {
    case TargetPlatform.android:
    case TargetPlatform.iOS:
      return true;
    default:
      return false;
  }
}

int get _hierHeaderHeight =>
    _hierTouchHeader ? _hierMobileHeaderHeight : _hierDesktopHeaderHeight;

Future<void> loadHierarchyJsonDemo(VolvoxGridController controller) async {
  final rawJson = utf8.decode(await controller.getDemoData('hierarchy'));
  final rows = (jsonDecode(rawJson) as List<dynamic>)
      .map((value) => Map<String, dynamic>.from(value as Map))
      .toList();
  final levels = _hierarchyLevels(rows);
  final types = rows.map((row) => _hierarchyString(row, 'Type')).toList();
  final visibleRows = rows.map(_hierarchyVisibleRow).toList();
  final sanitized = Uint8List.fromList(
    utf8.encode(jsonEncode(visibleRows)),
  );
  await controller.setColCount(_hierarchyColWidths.length);
  await controller.defineColumns(_hierarchyDefineColumnsRequest());
  final result = await controller.loadData(
    sanitized,
    options: (LoadDataOptions()..autoCreateColumns = false),
  );
  if (result.status == LoadDataStatus.LOAD_FAILED) {
    throw StateError('LoadData failed for embedded hierarchy demo');
  }

  await controller.configure(
    _hierarchyThemeConfig(
      _hierarchyMaxVisualDepth(levels),
      _hierarchyMaxLevel(levels),
    ),
  );

  final actionStyle = CellStyle()..foreground = _hierActionTextColor;
  final folderStyle = CellStyle()
    ..foreground = _hierFolderTextColor
    ..font = (Font()..bold = true);

  for (var row = 0; row < levels.length; row += 1) {
    final isFolder = row < types.length && types[row] == 'Folder';
    await controller.setRowOutlineLevel(row, levels[row]);
    await controller.setCellStyleRange(
      row,
      hierarchyActionColumn,
      row,
      hierarchyActionColumn,
      actionStyle,
    );
    if (isFolder) {
      await controller.setCellStyleRange(
        row,
        hierarchyNameColumn,
        row,
        hierarchyNameColumn,
        folderStyle,
      );
    }
  }
}

String _hierarchyString(Map<String, dynamic> row, String key) =>
    row[key]?.toString() ?? '';

Map<String, dynamic> _hierarchyVisibleRow(Map<String, dynamic> row) {
  return {
    for (final key in _hierarchyKeys) key: row[key] ?? '',
  };
}

List<int> _hierarchyLevels(List<Map<String, dynamic>> rows) {
  final rowsById = <String, Map<String, dynamic>>{};
  for (final row in rows) {
    final id = _hierarchyString(row, 'Id');
    if (id.isEmpty) {
      throw StateError('Hierarchy demo row is missing Id');
    }
    rowsById[id] = row;
  }

  final cache = <String, int>{};
  int depthOf(Map<String, dynamic> row, Set<String> visiting) {
    final id = _hierarchyString(row, 'Id');
    final cached = cache[id];
    if (cached != null) {
      return cached;
    }
    if (!visiting.add(id)) {
      throw StateError('Hierarchy demo data contains a parent cycle at $id');
    }
    final parentId = row['ParentId']?.toString() ?? '';
    var depth = 0;
    if (parentId.isNotEmpty) {
      final parent = rowsById[parentId];
      if (parent == null) {
        throw StateError(
            'Hierarchy demo data references missing parent $parentId');
      }
      depth = depthOf(parent, visiting) + 1;
    }
    visiting.remove(id);
    cache[id] = depth;
    return depth;
  }

  return rows.map((row) => depthOf(row, <String>{})).toList();
}

DefineColumnsRequest _hierarchyDefineColumnsRequest() {
  final request = DefineColumnsRequest();
  for (var col = 0; col < _hierarchyColWidths.length; col += 1) {
    final def = ColumnDef()
      ..index = col
      ..caption = _hierarchyCaptions[col]
      ..key = _hierarchyKeys[col]
      ..width = _hierarchyColWidths[col];
    if (col == _hierarchySizeColumn) {
      def.align = Align.ALIGN_RIGHT_CENTER;
    } else if (col == _hierarchyModifiedColumn) {
      def.dataType = ColumnDataType.COLUMN_DATA_DATE;
      def.format = _hierShortDateFormat;
    } else if (col == _hierarchyPermissionsColumn ||
        col == hierarchyActionColumn) {
      def.align = Align.ALIGN_CENTER_CENTER;
    }
    if (col == hierarchyActionColumn) {
      def.interaction = CellInteraction.CELL_INTERACTION_TEXT_LINK;
    }
    if (col == hierarchyNameColumn) {
      def.hidden = true;
    }
    request.columns.add(def);
  }
  return request;
}

int _hierarchyMaxVisualDepth(List<int> levels) {
  var hasMinLevel = false;
  var minLevel = 0;
  var maxLevel = 0;
  for (final level in levels) {
    if (level >= 0 && (!hasMinLevel || level < minLevel)) {
      hasMinLevel = true;
      minLevel = level;
    }
    if (level > maxLevel) {
      maxLevel = level;
    }
  }
  final depth = maxLevel - minLevel;
  return depth < 0 ? 0 : depth;
}

int _hierarchyMaxLevel(List<int> levels) {
  var hasMaxLevel = false;
  var maxLevel = 0;
  for (final level in levels) {
    if (level >= 0 && (!hasMaxLevel || level > maxLevel)) {
      hasMaxLevel = true;
      maxLevel = level;
    }
  }
  return hasMaxLevel ? maxLevel : 0;
}

int _hierarchyOutlineWidth(int maxOutlineDepth) {
  final sanitizedMaxDepth = maxOutlineDepth < 0 ? 0 : maxOutlineDepth;
  final width = (sanitizedMaxDepth + 1) * _hierOutlineIndent;
  return width < _hierMinOutlineIndicatorWidth
      ? _hierMinOutlineIndicatorWidth
      : width;
}

int _hierarchyExpanderWidth(int maxOutlineDepth) {
  return _hierarchyOutlineWidth(maxOutlineDepth) + _hierNameExpanderWidth;
}

GridConfig _hierarchyThemeConfig(int maxOutlineDepth, int maxOutlineLevel) {
  final outlineWidth = _hierarchyOutlineWidth(maxOutlineDepth);
  final expanderWidth = _hierarchyExpanderWidth(maxOutlineDepth);
  return GridConfig()
    ..themePreset = ThemePreset.THEME_AMBER
    ..layout = (LayoutConfig()..fixedRows = 0)
    ..selection = (SelectionConfig()..mode = SelectionMode.SELECTION_FREE)
    ..editing = (EditConfig()
      ..activation =
          (EditActivation()..trigger = EditTrigger.EDIT_TRIGGER_NONE))
    ..outline = (OutlineConfig()
      ..treeIndicator = TreeIndicatorStyle.TREE_INDICATOR_ARROWS_LEAF
      ..indicatorIndent = _hierOutlineIndent
      ..maxLevels = maxOutlineLevel < 0 ? 0 : maxOutlineLevel
      ..showLevelButtons = true
      ..labelColumn = hierarchyNameColumn
      ..treeColor = _hierTreeColor)
    ..interaction = (InteractionConfig()
      ..resize = (ResizePolicy()
        ..columns = true
        ..rows = false)
      ..autoSizeMouse = true
      ..headerFeatures = (HeaderFeatures()
        ..sort = false
        ..reorder = false
        ..chooser = false))
    ..indicators = (IndicatorsConfig()
      ..rowStart = (RowIndicatorConfig()
        ..visible = true
        ..width = expanderWidth
        ..autoSize = false
        ..allowResize = true
        ..slots.add(RowIndicatorSlot()
          ..kind = RowIndicatorSlotKind.ROW_INDICATOR_SLOT_EXPANDER
          ..width = expanderWidth
          ..visible = true))
      ..cornerTopStart = (CornerIndicatorConfig()
        ..visible = true
        ..slots.add(CornerIndicatorSlot()
          ..kind = CornerIndicatorSlotKind.CORNER_SLOT_OUTLINE_LEVELS
          ..width = outlineWidth
          ..visible = true))
      ..colTop = (ColIndicatorConfig()
        ..visible = true
        ..defaultRowHeight = _hierHeaderHeight
        ..bandRows = _hierHeaderBandRows
        ..cellModes = (ColIndicatorCellModes()
          ..modes.add(ColIndicatorCellMode.COL_INDICATOR_CELL_HEADER_TEXT))
        ..allowResize = true)
      ..appearance = IndicatorAppearance.INDICATOR_APPEARANCE_MODERN);
}
